File: data-structures/queues/visual-basic/queue_priority_OOP.vb
' Raspberry Pi Foundation
' Usage licensed under CC BY-SA 4
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


' A class for a node in a linked list
Class Node
    Private data As String
    Private priority As Integer
    Private nextItem As Node

    ' Constructor method
    Public Sub New(ByVal itemData As String, ByVal itemPriority As Integer)
        data = itemData
        priority = itemPriority
    End Sub


    Public Function GetData() As String
        Return data
    End Function


    Public Function GetPriority() As Integer
        Return priority
    End Function


    Public Function GetNext() As Node
        Return nextItem
    End Function


    Public Sub SetNext(ByVal newNext As Node)
        nextItem = newNext
    End Sub


    ' Testing
    Public Sub OutputNode()
        Dim nextData = "null"
        If nextItem IsNot Nothing Then
            nextData = nextItem.data
        End If
        Console.WriteLine($"Data: {data}, priority: {priority}, next: {nextData}")
    End Sub


End Class


' A priority queue stored as a linked list
Class PriorityQueue
    Private front As Node
    Private rear As Node

    Public Function GetFront() As Node
        Return front
    End Function


    Public Sub SetFront(ByVal newFront As Node)
        front = newFront
    End Sub


    Public Function GetRear() As Node
        Return rear
    End Function


    Public Sub SetRear(ByVal newRear As Node)
        rear = newRear
    End Sub


    ' Check if the queue is empty
    Public Function IsEmpty() As Boolean
        If front Is Nothing Then
            Return True
        Else
            Return False
        End If
    End Function


    ' Enqueue an item based on the priority
    Public Sub Enqueue(ByVal data As String, ByVal itemPriority As Integer)
        Dim newNode As Node = New Node(data, itemPriority)

        ' Check if the queue is empty
        If IsEmpty() Then
            front = newNode
            rear = newNode
        Else
            ' Check if new node should be the new front of queue
            If newNode.GetPriority() > front.GetPriority() Then
                ' Change the current head pointer
                newNode.SetNext(front)
                front = newNode
                ' Check if new node should be added to the end of the queue
            ElseIf newNode.GetPriority() <= rear.GetPriority() Then
                rear.SetNext(newNode)
                rear = newNode
            Else
                ' Find the insertion point
                Dim current = front
                Dim previous As Node = Nothing
                While current.GetPriority() >= newNode.GetPriority()
                    previous = current
                    current = current.GetNext()
                End While
                newNode.SetNext(current)
                previous.SetNext(newNode)
            End If
        End If
    End Sub


    ' Dequeue an item
    Public Function Dequeue() As String
        Dim dequeuedItem As String

        If IsEmpty() Then
            Console.WriteLine(vbLf & "Queue is empty - nothing to dequeue")
            dequeuedItem = ""
        Else
            ' Get data and change pointer to next item
            dequeuedItem = front.GetData()
            front = front.GetNext()
        End If

        Return dequeuedItem
    End Function


    ' Output the state of the queue
    Public Sub OutputQueue()
        ' Testing
        If front IsNot Nothing Then
            Console.WriteLine("------ State of the queue ------")
            Console.Write("Head ")
            Dim currentNode = front
            currentNode.OutputNode()
            While currentNode.GetNext() IsNot Nothing
                currentNode = currentNode.GetNext()
                currentNode.OutputNode()
            End While
            Console.WriteLine()
        End If
    End Sub


End Class


Module Program
    ' The Main method is the default entry point for all VB programs
    Sub Main()
        Console.WriteLine("### Priority queue (OOP) ###" & vbLf)

        ' Initialise the queue
        Dim myQueue As PriorityQueue = New PriorityQueue()

        ' Insert test data into the queue
        TestEnqueue(myQueue)

        ' Remove test data from the queue
        TestDequeue(myQueue)
    End Sub

    Sub TestEnqueue(ByVal myQueue As PriorityQueue)
        Console.WriteLine("Adding A with priority 3")
        myQueue.Enqueue("A", 3)
        myQueue.OutputQueue()

        Console.WriteLine("Adding B with priority 2 - should be behind A")
        myQueue.Enqueue("B", 2)
        myQueue.OutputQueue()

        Console.WriteLine("Adding C with priority 3 - should be in front of B")
        myQueue.Enqueue("C", 3)
        myQueue.OutputQueue()

        Console.WriteLine("Adding D with priority 3 - should be after C")
        myQueue.Enqueue("D", 3)
        myQueue.OutputQueue()

        Console.WriteLine("Adding E with priority 3 - should be after D")
        myQueue.Enqueue("E", 3)
        myQueue.OutputQueue()

        Console.WriteLine("Adding F with priority 1 - should be after B")
        myQueue.Enqueue("F", 1)
        myQueue.OutputQueue()

        Console.WriteLine("Adding G with priority 4 - should be head of queue")
        myQueue.Enqueue("G", 4)
        myQueue.OutputQueue()

        Console.WriteLine("Adding H with priority 5 - should be head of queue")
        myQueue.Enqueue("H", 5)
        myQueue.OutputQueue()

        Console.WriteLine("Adding I with priority 5 - should be after H")
        myQueue.Enqueue("I", 5)
        myQueue.OutputQueue()
    End Sub


    Sub TestDequeue(ByVal myQueue As PriorityQueue)
        Console.WriteLine(vbLf & "------ Dequeuing 10 items ------" & vbLf)

        For i = 0 To 9
            myQueue.Dequeue()
            myQueue.OutputQueue()
        Next
    End Sub


End Module

File: data-structures/queues/java/queue_priority_OOP.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

// A class for a node in a linked list
class Node {
    private String data;
    private int priority;
    private Node next;

    // Constructor method
    public Node(String itemData, int itemPriority) {
        data = itemData;
        priority = itemPriority;
    }

    public String getData() {
        return data;
    }

    public int getPriority() {
        return priority;
    }

    public Node getNext() {
        return next;
    }

    public void setNext(Node newNext) {
        next = newNext;
    }

    // Testing
    public void outputNode() {
        String nextData = "null";
        if (next != null)
            nextData = next.data;
        System.out.println("Data: " + data + ", priority: " + priority + ", next: " + nextData);
    }
}


// A priority queue stored as a linked list
class PriorityQueue {
    private Node front;
    private Node rear;

    public Node getFront() {
        return front;
    }

    public void setFront(Node newFront) {
        front = newFront;
    }

    public Node getRear() {
        return rear;
    }

    public void setRear(Node newRear) {
        rear = newRear;
    }

    // Check if the queue is empty
    public boolean isEmpty() {
        if (front == null)
            return true;
        else
            return false;
    }

    // Enqueue an item based on the priority
    public void enqueue(String data, int itemPriority) {
        Node newNode = new Node(data, itemPriority);

        // Check if the queue is empty
        if (isEmpty()) {
            front = newNode;
            rear = newNode;
        }
        else {
            // Check if new node should be the new front of queue
            if (newNode.getPriority() > front.getPriority()) {
                // Change the current head pointer
                newNode.setNext(front);
                front = newNode;
            }
            // Check if new node should be added to the end of the queue
            else if (newNode.getPriority() <= rear.getPriority()) {
                rear.setNext(newNode);
                rear = newNode;
            }
            else {
                // Find the insertion point
                Node current = front;
                Node previous = null;
                while (current.getPriority() >= newNode.getPriority()) {
                    previous = current;
                    current = current.getNext();
                }
                newNode.setNext(current);
                previous.setNext(newNode);
            }
        }
    }

    // Dequeue an item
    public String dequeue() {
        String dequeuedItem;

        if (isEmpty()) {
            System.out.println("\nQueue is empty - nothing to dequeue");
            dequeuedItem = "";
        }
        else {
            // Get data and change pointer to next item
            dequeuedItem = front.getData();
            front = front.getNext();
        }

        return dequeuedItem;
    }

    // Output the state of the queue
    public void outputQueue() {
        // Testing
        if (front != null) {
            System.out.println("------ State of the queue ------");
            System.out.print("Head ");
            Node currentNode = front;
            currentNode.outputNode();
            while (currentNode.getNext() != null) {
                currentNode = currentNode.getNext();
                currentNode.outputNode();
            }
            System.out.println();
        }
    }
}


class Queues {
    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        System.out.println("### Priority queue (OOP) ###\n");

        // Initialise the queue
        PriorityQueue myQueue = new PriorityQueue();

        // Insert test data into the queue
        testEnqueue(myQueue);

        // Remove test data from the queue
        testDequeue(myQueue);
    }

    public static void testEnqueue(PriorityQueue myQueue) {
        System.out.println("Adding A with priority 3");
        myQueue.enqueue("A", 3);
        myQueue.outputQueue();

        System.out.println("Adding B with priority 2 - should be behind A");
        myQueue.enqueue("B", 2);
        myQueue.outputQueue();

        System.out.println("Adding C with priority 3 - should be in front of B");
        myQueue.enqueue("C", 3);
        myQueue.outputQueue();

        System.out.println("Adding D with priority 3 - should be after C");
        myQueue.enqueue("D", 3);
        myQueue.outputQueue();

        System.out.println("Adding E with priority 3 - should be after D");
        myQueue.enqueue("E", 3);
        myQueue.outputQueue();

        System.out.println("Adding F with priority 1 - should be after B");
        myQueue.enqueue("F", 1);
        myQueue.outputQueue();

        System.out.println("Adding G with priority 4 - should be head of queue");
        myQueue.enqueue("G", 4);
        myQueue.outputQueue();

        System.out.println("Adding H with priority 5 - should be head of queue");
        myQueue.enqueue("H", 5);
        myQueue.outputQueue();

        System.out.println("Adding I with priority 5 - should be after H");
        myQueue.enqueue("I", 5);
        myQueue.outputQueue();
    }

    public static void testDequeue(PriorityQueue myQueue) {
        System.out.println("\n------ Dequeuing 10 items ------\n");

        for (int i = 0; i < 10; i++) {
            myQueue.dequeue();
            myQueue.outputQueue();
        }
    }


}